controllers/stardoginstance_controller_test.go

```diff
@@ -0,0 +1,457 @@
+				reconciliationContext: &ReconciliationContext{
+					context:       context.Background(),
+					conditions:    make(map[v1alpha1.StardogConditionType]v1alpha1.StardogCondition),
+					namespace:     namespace,
+					stardogClient: nil,
+				},
+				resource: createStardogInstance(namespace, stardogInstanceAltName, secretName, serverURL),
+			},
+			expectedFinalizers: []string{},
+			err: errors.New(fmt.Sprintf("cannot add finalizers for the StardogInstance %s/%s: "+
+				"stardoginstances.stardog.vshn.ch \"%s\" not found", namespace, stardogInstanceAltName, stardogInstanceAltName)),
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			fakeKubeClient, err := createKubeFakeClient(&tt.stardogInstance)
+			assert.NoError(t, err)
+			r := StardogInstanceReconciler{
+				Log:               testing2.TestLogger{},
+				ReconcileInterval: time.Duration(1),
+				Scheme:            scheme.Scheme,
+				Client:            fakeKubeClient,
+			}
+			err = r.addDeletionFinalizers(&tt.sir)
+
+			actualInstance := v1alpha1.StardogInstance{}
+			_ = fakeKubeClient.Get(context.Background(), types.NamespacedName{
+				Namespace: namespace,
+				Name:      stardogInstanceName,
+			}, &actualInstance)
+
+			assert.Equal(t, tt.err, err)
+			assert.ElementsMatch(t, actualInstance.GetFinalizers(), tt.expectedFinalizers)
+		})
+	}
+}
+
+func Test_deleteStardogInstance(t *testing.T) {
+
+	namespace := "namespace-test"
+	stardogInstanceName := "instance-test"
+	secretName := "secret-test"
+	serverURL := "url-test"
+	username := "admin"
+	password := "1234"
+
+	err := v1alpha1.AddToScheme(scheme.Scheme)
+	assert.NoError(t, err)
+
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+	stardogClient := stardogrestapi.NewMockExtendedBaseClientAPI(mockCtrl)
+
+	tests := []struct {
+		name               string
+		stardogInstance    v1alpha1.StardogInstance
+		sir                StardogInstanceReconciliation
+		secret             v1.Secret
+		expectedFinalizers []string
+		err                error
+	}{
+		{
+			name:            "GivenReconciliationContext_WhenFinalizersArePresent_ThenDeleteStardogInstance",
+			stardogInstance: *createStardogInstanceWithFinalizers(namespace, stardogInstanceName, secretName, serverURL),
+			secret:          *createFullSecret(namespace, secretName, username, password),
+			sir: StardogInstanceReconciliation{
+				reconciliationContext: &ReconciliationContext{
+					context:       context.Background(),
+					conditions:    make(map[v1alpha1.StardogConditionType]v1alpha1.StardogCondition),
+					namespace:     namespace,
+					stardogClient: stardogClient,
+				},
+				resource: createStardogInstanceWithFinalizers(namespace, stardogInstanceName, secretName, serverURL),
+			},
+			expectedFinalizers: []string{},
+			err:                nil,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			fakeKubeClient, err := createKubeFakeClient(&tt.stardogInstance, &tt.secret)
```

Comment:
I would suggest to do that in the next ticket, as you can see this PR is already too big to introduce other logic and/or other types of tests

Reply:
It's just code, so it's not too late to split PRs or so. But I don't have hard feelings, if the fake client covers your test case sufficiently, I'm fine as long as you don't request my help in a finalizer troubleshooting session :P

Reply:
There's not much time left unfortunately to deliver the operator and I lack the experience to add these kind of tests(I will have to research again, especially when the Stardog DB is involved). It would be really nice to have these e2e tests together with StardogDB.